### PR TITLE
Add new quoter that records gas, sqrt price after, and initialized ticks crossed.

### PR DESCRIPTION
**Pull request by willpote**

### CODE REVIEW

File: contracts/interfaces/IQuoterWithGas.sol [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+pragma solidity >=0.7.5;
+pragma abicoder v2;
+
+/// @title Quoter Interface
+/// @notice Supports quoting the calculated amounts from exact input or exact output swaps
+/// @notice For each pool also tells you the number of initialized ticks crossed and the sqrt price of the pool after the swap.
+/// @notice Also provides an estimate of the gas used by the swap.
+/// @dev These functions are not marked view because they rely on calling non-view functions and reverting
+/// to compute the result. They are also not gas efficient and should not be called on-chain.
+interface IQuoterWithGas {
```
Comment (moodysalem) [inline contracts/interfaces/IQuoterWithGas.sol:11]: i'm kinda anti using this, would rather delete, at least in part b/c the implementation includes the gas used in computing how many initialized ticks were crossed

File: contracts/interfaces/IQuoter2.sol [hunk 2]
```diff
@@ -0,0 +1,90 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+pragma solidity >=0.7.5;
+pragma abicoder v2;
+
+/// @title Quoter2 Interface
+/// @notice Supports quoting the calculated amounts from exact input or exact output swaps.
+/// @notice For each pool also tells you the number of initialized ticks crossed and the sqrt price of the pool after the swap.
+/// @dev These functions are not marked view because they rely on calling non-view functions and reverting
+/// to compute the result. They are also not gas efficient and should not be called on-chain.
+interface IQuoter2 {
```
Comment (moodysalem) [inline contracts/interfaces/IQuoter2.sol:10]: can we do `QuoterV2`?

**Review by moodysalem — CHANGES_REQUESTED** [review]

File: contracts/interfaces/IQuoterWithGas.sol (hunk 1 above)
Comment (willpote) [reply contracts/interfaces/IQuoterWithGas.sol]: Moved to within Quoter. As discussed we may end up not using it but lets see

File: contracts/lens/Quoter.sol [hunk 3]
```diff
@@ -66,7 +66,7 @@ contract Quoter is IQuoter, IUniswapV3SwapCallback, PeripheryImmutableState {
     }
 
     /// @dev Parses a revert reason that should contain the numeric quote
-    function parseRevertReason(bytes memory reason) private pure returns (uint256) {
+    function parseRevertReason(bytes memory reason) private view returns (uint256) {
```
[the remark this answers is not in the document]
Comment (willpote) [reply contracts/lens/Quoter.sol]: Fixed

File: contracts/interfaces/IQuoter2.sol (hunk 2 above)
Comment (willpote) [reply contracts/interfaces/IQuoter2.sol]: Done

File: contracts/lens/QuoterV2.sol [hunk 4]
```diff
@@ -0,0 +1,332 @@
...
+        int24 tickAfter;
+        (, tickBefore, , , , , ) = pool.slot0();
+        (amount, sqrtPriceX96After, tickAfter) = parseRevertReason(reason);
+
+        initializedTicksCrossed = countInitializedBitsCrossed(pool, tickBefore, tickAfter);
+
+        return (amount, sqrtPriceX96After, initializedTicksCrossed, gasEstimate);
+    }
+
+    function countInitializedBitsCrossed(
```
Comment (moodysalem) [inline contracts/lens/QuoterV2.sol:130]: this method is really complex--it probably should be well tested as a library function to know that it works, given there are many edge cases

File: contracts/lens/QuoterV2.sol [hunk 5]
```diff
@@ -0,0 +1,332 @@
...
+                // ending tick.
+                if (wordPos == wordPosAfter) {
+                    mask = mask & (~(uint256(0)) << bitPosAfter);
+                }
+
+                uint256 masked = pool.tickBitmap(wordPos) & mask;
+                initializedTicksCrossed += countBits(masked);
+                wordPos--;
+                // Reset our mask so we consider all bits on the next iteration.
+                mask = ~(uint256(0));
```
Comment (moodysalem) [inline contracts/lens/QuoterV2.sol:175]: ditto

File: contracts/lens/QuoterV2.sol [hunk 6]
```diff
@@ -0,0 +1,332 @@
...
+            }
+        } else {
+            // Our first mask should include the starting tick, and everything to its right.
+            uint256 mask = (1 << (bitPos + 1)) - 1;
+            while (wordPos >= wordPosAfter) {
+                // If we're on the final tick bitmap page, ensure we only count up to our
+                // ending tick.
+                if (wordPos == wordPosAfter) {
+                    mask = mask & (~(uint256(0)) << bitPosAfter);
```
Comment (moodysalem) [inline contracts/lens/QuoterV2.sol:168]: ditto

**Review by moodysalem — CHANGES_REQUESTED** [review]

File: contracts/test/PoolTicksHelperTest.sol [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+
+pragma solidity >=0.6.0;
+
+import '../libraries/PoolTicksHelper.sol';
+
+contract PoolTicksHelperTest {
+    using PoolTicksHelper for IUniswapV3Pool;
+    IUniswapV3Pool public pool;
+
+    function setPool(IUniswapV3Pool _pool) external {
```
Comment (moodysalem) [inline contracts/test/PoolTicksHelperTest.sol:12]: you could just take this as an additional argument to countInitializedBitsCrossed

File: package.json [hunk 8]
```diff
@@ -33,9 +33,11 @@
     "@uniswap/lib": "^4.0.1-alpha",
     "@uniswap/v2-core": "1.0.1",
     "@uniswap/v3-core": "1.0.0",
-    "base64-sol": "1.0.1"
+    "base64-sol": "1.0.1",
+    "hardhat-watcher": "^2.1.1"
   },
   "devDependencies": {
+    "@nomiclabs/ethereumjs-vm": "^4.2.2",
```
Comment (moodysalem) [inline package.json:40]: is this old?

File: contracts/libraries/PoolTicksHelper.sol [hunk 9]
```diff
@@ -0,0 +1,78 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+import '@uniswap/v3-core/contracts/interfaces/IUniswapV3Pool.sol';
+
+pragma solidity >=0.6.0;
+
+
+library PoolTicksHelper {
```
Comment (moodysalem) [inline contracts/libraries/PoolTicksHelper.sol:7]: helper is an overloaded/unhelpful word. `Counter` or no suffix preferred, also some natspec on the library maybe?

File: test/PoolTicksHelper.spec.ts [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+
+    it('counts all ticks in a page', async () => {
+      await pool.mock.tickBitmap.withArgs(0).returns(ethers.constants.MaxUint256); // 1100
+      await pool.mock.tickBitmap.withArgs(1).returns(0x0); 
+      const result = await poolTicksHelper.countInitializedBitsCrossed(bitIdxToTick(0), bitIdxToTick(255, 1));
+      expect(result).to.be.eq(256);
+    })
+
+    it('counts ticks to left of start and right of end on same page', async () => {
+      await pool.mock.tickBitmap.withArgs(0).returns(0xf10f); // 1111000100001111
```
Comment (moodysalem) [inline test/PoolTicksHelper.spec.ts:67]: you can use `0b1111000100001111` according to the chrome console

File: test/PoolTicksHelper.spec.ts [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+  const bitIdxToTick = (idx: number, page = 0) => {
+    return (idx * TICK_SPACING) + (page * 256 * TICK_SPACING);
+  }
+  
+  before(async () => {
+    PoolAbi = await artifacts.readArtifact('IUniswapV3Pool')
+    const poolTicksHelperFactory = await ethers.getContractFactory('PoolTicksHelperTest')
+    poolTicksHelper = (await poolTicksHelperFactory.deploy()) as PoolTicksHelperTest
+    pool = await deployMockContract(wallets[0], PoolAbi.abi)
+    await pool.mock.tickSpacing.returns(TICK_SPACING);
```
Comment (moodysalem) [inline test/PoolTicksHelper.spec.ts:29]: is it important to try different tick spacings?

**Review by moodysalem — APPROVED** [review]

**Comment by moodysalem** [comment]
also need to update yarn.lock to get the ci to run

File: contracts/libraries/PoolTicksCounter.sol [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+import '@uniswap/v3-core/contracts/interfaces/IUniswapV3Pool.sol';
+
+pragma solidity >=0.6.0;
+
+library PoolTicksCounter {
+    /// @dev This function counts the number of initialized ticks 't' such that tickBefore <= t < tickAfter.
+    /// Therefore if tickAfter happens to be an initialized tick it is *not* counted. This is because the
+    /// function is intended to be used for gas estimation and the cost associated with an intialized tick
+    /// only occurs once the initialized tick is passed.
+    function countInitializedBitsCrossed(
```
Comment (moodysalem) [inline contracts/libraries/PoolTicksCounter.sol:11]: `countInitializedBitsCrossed` -> `countInitializedTicksCrossed`

File: contracts/libraries/PoolTicksCounter.sol [hunk 13]
```diff
@@ -0,0 +1,81 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+import '@uniswap/v3-core/contracts/interfaces/IUniswapV3Pool.sol';
+
+pragma solidity >=0.6.0;
+
+library PoolTicksCounter {
+    /// @dev This function counts the number of initialized ticks 't' such that tickBefore <= t < tickAfter.
+    /// Therefore if tickAfter happens to be an initialized tick it is *not* counted. This is because the
+    /// function is intended to be used for gas estimation and the cost associated with an intialized tick
+    /// only occurs once the initialized tick is passed.
```
Comment (moodysalem) [inline contracts/libraries/PoolTicksCounter.sol:10]: this statement isn't clear to me

> This function counts the number of initialized ticks 't' such that tickBefore <= t < tickAfter.

tickBefore and tickAfter are absolute ticks, but t is a count of ticks crossed, so this relation is confusing

i think i know what you're trying to say, but not sure how to better say it.. you're counting ticks that are initialized and have indices <= tickBefore and > tickAfter

also <= and > depends on direction no?

**Review by moodysalem — APPROVED** [review]

**Pull request merged**